### PR TITLE
[chore] - chore: store fragment content type

### LINKED ISSUE #5417: Store content type alongside content fragment
Content fragments were initially designed to support only Slack threads or file attachments. To prepare for Vision, we need to store images as well, including their content type, so they can be properly displayed in the UI.

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description

This PR aims at storing the content type within a fragment. 

To do so we retrieve the [MIME type](https://developer.mozilla.org/en-US/docs/Glossary/MIME_type) from the uploaded `File` object and store it in `contentType`. Formerly all previous uploaded file had a "file_attachment" contentType. To achieve this we introduce a `supportedContentFragment` type, listing all the MIME types currently supported. This allows us to control assert if the uploaded content fragment is supported when uploading it.

Note that for slack content fragment we create a contentType: `dust-application/slack` (formerly "slack_thread")

References:
- Task: https://github.com/dust-tt/dust/issues/5417

## Risk

Moderated, we need to make sure that we are backward compatible while backfilling data.

## Deploy Plan

1. Deploy `front`
2. Deploy `connectors`

### CODE REVIEW

File: front/components/assistant/conversation/ContentFragment.tsx [hunk 1]
```diff
@@ -6,8 +6,16 @@ export function ContentFragment({ message }: { message: ContentFragmentType }) {
   let logoType: "document" | "slack" = "document";
   switch (message.contentType) {
     case "slack_thread_content":
+    case "dust-application/slack":
       logoType = "slack";
       break;
+    case "text/plain":
```
Comment (flvndvd) [inline front/components/assistant/conversation/ContentFragment.tsx:12]: We can probably do `.startsWith("text/")` or move this to a common helper.

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 2]
```diff
@@ -192,6 +198,15 @@ export function AssistantInputBar({
                   "PDF uploads are limited to 100Mb per file. Please consider uploading a smaller file.",
               });
               return;
+            } else if (isSupportedContentFormat(file.type)) {
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:201]: I guess you want the not (`!`)?

File: front/components/assistant/conversation/input_bar/InputBarContainer.tsx [hunk 3]
```diff
@@ -26,6 +26,8 @@ export const INPUT_BAR_ACTIONS = ["attachment", "quick-actions"] as const;
 
 export type InputBarAction = (typeof INPUT_BAR_ACTIONS)[number];
 
+const supportedFileExtensions = [".txt", ".csv", ".tsv", ".md", ".pdf"];
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBarContainer.tsx:29]: Why adding `.tsv`? Can we ensure that this PR is only about refactoring and does not improve coverage of content type?

File: front/components/assistant/conversation/lib.ts [hunk 4]
```diff
@@ -229,6 +231,34 @@ export async function createConversationWithMessage({
...
+  const validatedContentFragments: {
+    content: string;
+    title: string;
+    url: null;
+    contentType: ContentFragmentContentType;
+    context: { profilePictureUrl: string | null };
+  }[] = [];
+
+  for (const cf of contentFragments) {
+    if (isSupportedContentFormat(cf.file.type)) {
```
Comment (flvndvd) [inline front/components/assistant/conversation/lib.ts:243]: This seems to be a bit late, we already created the content fragment.

File: front/lib/resources/content_fragment_resource.ts [hunk 5]
```diff
@@ -252,3 +254,22 @@ export async function getContentFragmentText({
 
   return getPrivateUploadBucket().fetchFileContent(filePath);
 }
+
+export async function getSignedUrlForContentFragment({
```
Comment (flvndvd) [inline front/lib/resources/content_fragment_resource.ts:258]: This does not seem used, right?

File: types/src/front/api_handlers/internal/assistant.ts [hunk 6]
```diff
@@ -14,8 +14,16 @@ export const InternalPostContentFragmentRequestBodySchema = t.type({
   content: t.string,
   url: t.union([t.string, t.null]),
   contentType: t.union([
-    t.literal("slack_thread_content"),
+    t.literal("text/plain"),
```
Comment (flvndvd) [inline types/src/front/api_handlers/internal/assistant.ts:17]: Seems like a lot of duplication here...

File: front/components/assistant/conversation/lib.ts (hunk 4 above)
Comment (JulesBelveze) [reply front/components/assistant/conversation/lib.ts:243]: @flvndvd I would personally modify `ContentFragmentInput` to add an extra field `type` or something. This way we would ensure we actually support the type when creating a `ContentFragmentInput`. wdyt?

Comment (flvndvd) [reply front/components/assistant/conversation/lib.ts:243]: Seen IRL, you can create a function `uploadContentFragments` that returns the right type.

File: types/src/front/assistant/conversation.ts [hunk 7]
```diff
@@ -166,6 +183,23 @@ export function isContentFragmentType(
   return arg.type === "content_fragment";
 }
 
+export function isSupportedContentFormat(
```
Comment (flvndvd) [inline types/src/front/assistant/conversation.ts:186]:
```suggestion
export function isSupportedContentType(
```

File: types/src/front/assistant/conversation.ts [hunk 8]
```diff
@@ -166,6 +183,23 @@ export function isContentFragmentType(
...
+  format: unknown
+): format is ContentFragmentContentType {
+  return (
+    typeof format === "string" &&
+    supportedContentFragment.includes(format as ContentFragmentContentType)
+  );
+}
+
+export function isSupportedTextContentFormat(
```
Comment (flvndvd) [inline types/src/front/assistant/conversation.ts:195]: Align naming here as well.

File: types/src/front/api_handlers/public/assistant.ts [hunk 9]
```diff
@@ -37,8 +37,13 @@ export const PublicPostContentFragmentRequestBodySchema = t.type({
   content: t.string,
   url: t.union([t.string, t.null]),
   contentType: t.union([
-    t.literal("slack_thread_content"),
+    t.literal("text/plain"),
```
Comment (flvndvd) [inline types/src/front/api_handlers/public/assistant.ts:40]: You can infer this from the `supportedContentFragment`.

Comment (flvndvd) [reply types/src/front/api_handlers/public/assistant.ts:40]: See here: https://github.com/dust-tt/dust/blob/0d120635213c1436145ce4e6ad671bd237d32303/front/lib/api/pagination.ts#L21-L42.

File: front/lib/resources/content_fragment_resource.ts [hunk 10]
```diff
@@ -227,7 +229,7 @@ export async function storeContentFragmentText({
 
   await getPrivateUploadBucket().uploadRawContentToBucket({
     content,
-    contentType: "text/plain",
+    contentType: contentType,
```
Comment (flvndvd) [inline front/lib/resources/content_fragment_resource.ts:232]: `nit`
```suggestion
    contentType,
```

File: front/lib/resources/content_fragment_resource.ts [hunk 11]
```diff
@@ -208,11 +208,13 @@ export async function storeContentFragmentText({
   conversationId,
   messageId,
   content,
+  contentType,
 }: {
   workspaceId: string;
   conversationId: string;
   messageId: string;
   content: string;
+  contentType: string;
```
Comment (flvndvd) [inline front/lib/resources/content_fragment_resource.ts:217]: We could use proper typing here.

File: front/lib/api/assistant/conversation.ts [hunk 12]
```diff
@@ -1630,21 +1631,21 @@ export async function postNewContentFragment(
...
-  const sourceUrl =
-    contentType === "file_attachment"
-      ? fileAttachmentLocation({
-          workspaceId: owner.sId,
-          conversationId: conversation.sId,
-          messageId,
-          contentFormat: "raw",
-        }).downloadUrl
-      : url;
+  const sourceUrl = isSupportedTextContentFormat(contentType)
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:1634]: `isSupportedTextContentFormat` should not include `slack_thread`.

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 13]
```diff
@@ -392,3 +415,23 @@ export function FixedAssistantInputBar({
     </div>
   );
 }
+
+function isMarkdownFile(file: File): boolean {
+  if (file.type === "") {
+    const fileExtension = file.name.split(".").pop()?.toLowerCase();
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:421]: `nit`
```suggestion
    const fileExtension = file.name.split(".").at(-1)?.toLowerCase();
```

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 14]
```diff
@@ -392,3 +415,23 @@ export function FixedAssistantInputBar({
...
+    // Check if the file extension corresponds to a markdown file
+    return fileExtension === "md" || fileExtension === "markdown";
+  }
+  return file.type === "text/markdown";
+}
+
+function copyFileWithMimeType(originalFile: File, mimeType: string): File {
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:428]: Let's avoid doing some kind of copy of "file". We use APIs to create a file, let's not mess with this. Instead feel free to create a helper something like `getMimeTypeFromFile()` and use it when you want to access the `mimeType`, WDYT?

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 15]
```diff
@@ -192,7 +204,17 @@ export function AssistantInputBar({
                   "PDF uploads are limited to 100Mb per file. Please consider uploading a smaller file.",
               });
               return;
+            } else if (!isSupportedContentFormat(file.type)) {
+              const prettyfiedSupportedFormats =
+                supportedContentFragment.join(", ");
+              sendNotification({
+                type: "error",
+                title: "Error uploading file.",
+                description: `Unsupported file format, please use any of the following: ${prettyfiedSupportedFormats}`,
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:213]: Probably too much 🙃.

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 16]
```diff
@@ -178,10 +185,15 @@ export function AssistantInputBar({
         setIsProcessingContentFragment(true);
         try {
           let totalSize = 0;
-          for (const file of (event?.target as HTMLInputElement)?.files || []) {
+          for (let file of (event?.target as HTMLInputElement)?.files || []) {
             if (!file) {
               return;
             }
+            if (isMarkdownFile(file)) {
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:192]: Let's not expose this outside of a helper.

File: front/components/assistant/conversation/lib.ts [hunk 17]
```diff
@@ -40,6 +41,10 @@ export type ContentFragmentInput = {
   file: File;
 };
 
+export type UploadedContentFragment = ContentFragmentInput & {
```
Comment (flvndvd) [inline front/components/assistant/conversation/lib.ts:44]: Move to `@types`.

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 18]
```diff
@@ -192,7 +197,17 @@ export function AssistantInputBar({
                   "PDF uploads are limited to 100Mb per file. Please consider uploading a smaller file.",
               });
               return;
+            } else if (
+              !isSupportedContentFragmentType(getMimeTypeFromFile(file))
+            ) {
+              sendNotification({
+                type: "error",
+                title: "Error uploading file.",
+                description: `Unsupported file format.`,
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:206]:
```suggestion
                description: "Unsupported file format.",
```

File: front/lib/client/handle_file_upload.ts [hunk 19]
```diff
@@ -1,19 +1,40 @@
...
+import { getMimeTypeFromFile } from "@app/lib/file";
+PDFJS.GlobalWorkerOptions.workerSrc = `//cdnjs.cloudflare.com/ajax/libs/pdf.js/${PDFJS.version}/pdf.worker.mjs`;
 
-export async function handleFileUploadToText(
-  file: File
-): Promise<Result<{ title: string; content: string }, Error>> {
+export async function handleFileUploadToText(file: File): Promise<
+  Result<
+    {
+      title: string;
```
Comment (flvndvd) [inline front/lib/client/handle_file_upload.ts:16]: Any reason for not using `UploadedContentFragment` type?

File: front/components/assistant/conversation/input_bar/InputBar.tsx [hunk 20]
```diff
@@ -192,7 +197,17 @@ export function AssistantInputBar({
                   "PDF uploads are limited to 100Mb per file. Please consider uploading a smaller file.",
               });
               return;
+            } else if (
```
Comment (flvndvd) [inline front/components/assistant/conversation/input_bar/InputBar.tsx:200]: No need for it we can centralize the logic in `handleFileUploadToText`.

File: front/lib/client/handle_file_upload.ts [hunk 21]
```diff
@@ -1,19 +1,40 @@
...
+    },
+    Error
+  >
+> {
   return new Promise((resolve) => {
     const handleFileLoadedText = (e: ProgressEvent<FileReader>) => {
       const content = e.target?.result;
       if (content && typeof content === "string") {
-        return resolve(new Ok({ title: file.name, content }));
+        const contentType = getMimeTypeFromFile(file);
```
Comment (flvndvd) [inline front/lib/client/handle_file_upload.ts:27]: Let's extract it so we call it only once.

File: front/lib/client/handle_file_upload.ts [hunk 22]
```diff
@@ -64,16 +95,12 @@ export async function handleFileUploadToText(
     };
 
     try {
-      if (file.type === "application/pdf") {
+      const mimeType = getMimeTypeFromFile(file);
```
Comment (flvndvd) [inline front/lib/client/handle_file_upload.ts:98]: ✂️ If extracted above.

File: front/lib/resources/content_fragment_resource.ts [hunk 23]
```diff
@@ -208,11 +213,13 @@ export async function storeContentFragmentText({
   conversationId,
   messageId,
   content,
+  contentType,
 }: {
   workspaceId: string;
   conversationId: string;
   messageId: string;
   content: string;
+  contentType: SupportedContentFragmentType;
```
Comment (flvndvd) [inline front/lib/resources/content_fragment_resource.ts:222]: Let's keep `text/plain`.

File: types/src/front/api_handlers/internal/assistant.ts [hunk 24]
```diff
@@ -13,10 +15,7 @@ export const InternalPostContentFragmentRequestBodySchema = t.type({
   title: t.string,
   content: t.string,
   url: t.union([t.string, t.null]),
-  contentType: t.union([
-    t.literal("slack_thread_content"),
-    t.literal("file_attachment"),
-  ]),
+  contentType: getSupportedContentFragmentCodec(),
```
Comment (flvndvd) [inline types/src/front/api_handlers/internal/assistant.ts:18]:
```suggestion
  contentType: getSupportedContentFragmentTypeCodec(),
```

File: types/src/front/assistant/conversation.ts [hunk 25]
```diff
@@ -142,9 +144,36 @@ export type ContentFragmentContextType = {
   profilePictureUrl: string | null;
 };
 
-export type ContentFragmentContentType =
-  | "slack_thread_content"
-  | "file_attachment";
+// TODO (26/06/2024 jules): remove "slack_thread_content" and "file_attachment"
+// after backfilling data
+export const supportedTextContentFragment = [
+  "text/plain",
```
Comment (flvndvd) [inline types/src/front/assistant/conversation.ts:150]: You need all values from the former `isVisualText()` function:

```
    "text/plain",
    "text/csv",
    "text/markdown",
    "text/tsv",
    "text/comma-separated-values",
    "text/tab-separated-values",
```

File: types/src/front/assistant/conversation.ts [hunk 26]
```diff
@@ -142,9 +144,36 @@ export type ContentFragmentContextType = {
...
+export const supportedContentFragment = [
+  ...supportedTextContentFragment,
+  "slack_thread_content",
+  "dust-application/slack",
+] as const;
+
+export type SupportedContentFragmentType =
+  (typeof supportedContentFragment)[number];
+
+export function getSupportedContentFragmentCodec(): t.Mixed {
```
Comment (flvndvd) [inline types/src/front/assistant/conversation.ts:169]: Can we move all those types to a content fragment file?

File: front/lib/client/handle_file_upload.ts (hunk 19 above)
Comment (JulesBelveze) [reply front/lib/client/handle_file_upload.ts:16]: Initially we were not returning the `File` object. I sticked to this logic

**Review by flvndvd — APPROVED** [review]
LGTM.

**Pull request merged**